service/history/workflow/metrics.go

```diff
@@ -60,36 +60,40 @@ func emitMutableStateStatus(
-	metrics.TotalSignalCount.With(metricsHandler).Record(stats.TotalSignalCount)
-	metrics.BufferedEventsSize.With(metricsHandler).Record(int64(stats.BufferedEventsSize))
-	metrics.BufferedEventsCount.With(metricsHandler).Record(int64(stats.BufferedEventsCount))
+
+	batchHandler := metricsHandler.StartBatch("mutable_state_status")
+	defer batchHandler.Close()
+	metrics.MutableStateSize.With(batchHandler).Record(int64(stats.TotalSize))
+	metrics.ExecutionInfoSize.With(batchHandler).Record(int64(stats.ExecutionInfoSize))
+	metrics.ExecutionStateSize.With(batchHandler).Record(int64(stats.ExecutionStateSize))
+	metrics.ActivityInfoSize.With(batchHandler).Record(int64(stats.ActivityInfoSize))
+	metrics.ActivityInfoCount.With(batchHandler).Record(int64(stats.ActivityInfoCount))
+	metrics.TotalActivityCount.With(batchHandler).Record(stats.TotalActivityCount)
+	metrics.TimerInfoSize.With(batchHandler).Record(int64(stats.TimerInfoSize))
+	metrics.TimerInfoCount.With(batchHandler).Record(int64(stats.TimerInfoCount))
+	metrics.TotalUserTimerCount.With(batchHandler).Record(stats.TotalUserTimerCount)
+	metrics.ChildInfoSize.With(batchHandler).Record(int64(stats.ChildInfoSize))
+	metrics.ChildInfoCount.With(batchHandler).Record(int64(stats.ChildInfoCount))
+	metrics.TotalChildExecutionCount.With(batchHandler).Record(stats.TotalChildExecutionCount)
+	metrics.RequestCancelInfoSize.With(batchHandler).Record(int64(stats.RequestCancelInfoSize))
+	metrics.RequestCancelInfoCount.With(batchHandler).Record(int64(stats.RequestCancelInfoCount))
+	metrics.TotalRequestCancelExternalCount.With(batchHandler).Record(stats.TotalRequestCancelExternalCount)
+	metrics.SignalInfoSize.With(batchHandler).Record(int64(stats.SignalInfoSize))
+	metrics.SignalInfoCount.With(batchHandler).Record(int64(stats.SignalInfoCount))
+	metrics.TotalSignalExternalCount.With(batchHandler).Record(stats.TotalSignalExternalCount)
+	metrics.SignalRequestIDSize.With(batchHandler).Record(int64(stats.SignalRequestIDSize))
+	metrics.SignalRequestIDCount.With(batchHandler).Record(int64(stats.SignalRequestIDCount))
+	metrics.TotalSignalCount.With(batchHandler).Record(stats.TotalSignalCount)
+	metrics.BufferedEventsSize.With(batchHandler).Record(int64(stats.BufferedEventsSize))
+	metrics.BufferedEventsCount.With(batchHandler).Record(int64(stats.BufferedEventsCount))
 
 	if stats.HistoryStatistics != nil {
-		metrics.HistorySize.With(metricsHandler).Record(int64(stats.HistoryStatistics.SizeDiff))
-		metrics.HistoryCount.With(metricsHandler).Record(int64(stats.HistoryStatistics.CountDiff))
+		metrics.HistorySize.With(batchHandler).Record(int64(stats.HistoryStatistics.SizeDiff))
+		metrics.HistoryCount.With(batchHandler).Record(int64(stats.HistoryStatistics.CountDiff))
 	}
 
 	for category, taskCount := range stats.TaskCountByCategory {
+		// We use the metricsHandler rather than the batchHandler here because the same metric is repeatedly sent
```

Comment:
Can't we support this in the batch handler? This limits the approach a bit.

Reply:
We could add it to the batch and leave it up to the handler implementation to decide what to do with duplicate metrics/labels. The main issue with this particular line is it also adds extra labels. We can't add those labels to a batch as is because they couldn't be correlated with the specific task count and would apply to the batch as a whole. We could change the metric name/label to be `[categoryname]-task_count` but that would break existing task_count metrics (unless the concatenation happened in the handler but that seems like a bit of a hack).

Reply:
Thinking about this some more I'm going to add it to the batch and we can decide what to do about it in the underlying implementation. I think I can make some kind of nested structure to handle it.